src/cdomains/apron/affineEqualityDomain.apron.ml

```diff
@@ -0,0 +1,725 @@
+
+  let leq t1 t2 =
+    if is_bot t1 || is_top t2 then true else
+    if is_bot t2 || is_top t1 then false else (
+      Environment.equal t1.env t2.env &&
+      match t1.d, t2.d with
+      | Some x, Some y -> begin match Matrix.normalize @@ Matrix.append_matrices x y with
+          | Some m -> Matrix.equal m x
+          | None -> false
+        end
+      | _ -> false )
+
+  let leq t1 t2 =
+    let res = leq t1 t2 in
+    if M.tracing then M.tracel "leq" "leq a: %s b: %s -> %b \n" (show t1) (show t2) res ;
+    res
+
+  let join a b =
+    let rec lin_disjunc r s a b =
+      if s >= Matrix.num_cols a then a else
+        let case_two a r col_b =
+          let col_b = let a_length, b_length = Matrix.num_rows a, Vector.length col_b in
+            match Int.compare a_length b_length with
+            | -1 -> Vector.keep_vals col_b a_length
+            |  1 -> Vector.append col_b @@ Vector.zero_vec (a_length - b_length)
+            | _ -> col_b
+          in
+          let a_r = Matrix.get_row a r in
+          let mapping = Matrix.map2i (fun i x y -> if i < r then
+                                         Vector.map2 (+:) x (Vector.apply_with_c ( *:) y a_r) else x ) a col_b
+          in Matrix.remove_row mapping r
+        in
+        let case_three a b col_a col_b max =
+          let col_a, col_b = Vector.keep_vals col_a max, Vector.keep_vals col_b max in
+          if Vector.equal col_a col_b then (a, b, max) else
+            let a_rev, b_rev = Vector.rev col_a,  Vector.rev col_b in
+            let i, (x, y) = Vector.find2i (fun i x y -> x <> y) a_rev b_rev in
+            let r, diff = Vector.length a_rev - (i + 1), x -: y  in
+            let a_r, b_r = Matrix.get_row a r, Matrix.get_row b r in
+            let multiply_by_t' m col_v t i =
+              let zero_vec = Vector.of_list @@ List.init (Matrix.num_rows m - Vector.length col_v) (fun x -> of_int 0) in
+              let cs = Vector.append col_v zero_vec in
+              Matrix.map2i (fun i' x c -> if i' <= i then let beta = c /: diff in
+                               let mul_t = Vector.apply_with_c ( *:) beta t in Vector.map2 (-:) x mul_t else x) m cs
+            in
+            let sub_col = Vector.map2 (fun x y -> x -: y) col_a col_b in
+            Matrix.remove_row (multiply_by_t' a sub_col a_r max) r, Matrix.remove_row (multiply_by_t' b sub_col b_r max) r, (max - 1)
+        in
+        let col_a, col_b = Matrix.get_col a s, Matrix.get_col b s in
+        match Int.compare (Matrix.num_rows a) r, Int.compare (Matrix.num_rows b) r with
+        | 1 , 1 ->  let a_rs, b_rs = Mpqf.to_float (Vector.nth col_a r), Mpqf.to_float (Vector.nth col_b r) in
+          begin match  a_rs, b_rs with
+            | 1., 1. -> lin_disjunc (r + 1) (s + 1) a b
+            | 1., 0. -> lin_disjunc r (s + 1) (case_two a r col_b) b
+            | 0., 1. -> lin_disjunc r (s + 1) a (case_two b r col_a)
+            | 0., 0. ->  let new_a, new_b, new_r = case_three a b col_a col_b r in
```

Comment:
This match involves floating-point equality of _non-constant_ numbers, which is very fragile.